lib/ansible/modules/pip.py

```diff
@@ -611,6 +614,62 @@ def setup_virtualenv(module, env, chdir, out, err):
     return out, err, cmd
 
 
+def _normalize_vcs_packages(
+        module: AnsibleModule,
+        package_list: list[Package],
+        pip: list[str]
+) -> list[Package]:
+    """Return a list of packages with names extracted from metadata in VCS"""
+    vcs_packages = [pkg_str for pkg in package_list if _is_vcs_url(pkg_str := str(pkg))]
+
+    if not vcs_packages:
+        return package_list
+
+    # Unsetting 'FORCE_COLOR' with 'NO_COLOR' and TTY_COMPATIBLE settings, helps to
+    # prevent color codes from being injected into and corrupting the JSON output.
+    os.environ.pop('FORCE_COLOR', None)
+
+    rc, json_out, err = module.run_command(
+        [*pip, 'install', '--dry-run', '--ignore-installed', '--quiet', '--report=-', *vcs_packages],
+        environ_update={
+            'NO_COLOR': '1',
+            'TTY_COMPATIBLE': '0',
+        },
+    )
+
+    if rc == 0:
+        # If it succeeds, handle the report
+        report = json.loads(json_out)
+        package_objects = (Package(install_report['metadata']['name'], version_string=install_report['metadata']['version'])
+                           for install_report in report['install'])
+    else:
+        # Else, if that fails due to --dry-run not being present in pip
+        # versions older than 22.2 (ex: Ubuntu 22.04), fall back to using
+        # the potentially faulty pip download method and warn the user.
+
+        pip_downloads_dir = Path(module.tmpdir) / 'pipdownloads'
+        if pip_downloads_dir.exists():
+            shutil.rmtree(pip_downloads_dir)
+            pip_downloads_dir.mkdir()
+
+        rc, out, err = module.run_command([*pip, 'download', f'--dest={pip_downloads_dir}', '--no-deps', *vcs_packages])
+
+        if rc != 0:
+            # If it fails (usually due to no pkg at url), just dump the error
+            module.fail_json(rc=rc, msg=out, err=err)
+
+        # Only from this point onward can we be sure that we're using the fallback, warn now.
+        module.warn("Using check_mode with vcs packages is potentially error prone on pip versions <22.2")
```

Comment:
I imagine this goes beyond VCS and would touch any absolute source identifier, including local file/folder paths, tarball URLs and VCS pointers, right?
```suggestion
        module.warn("Using check mode with URL-sourced packages is potentially error-prone on pip versions older than 22.2")
```